beacon-chain/core/epoch/epoch_processing.go

```diff
@@ -191,3 +192,89 @@ func ProcessEjections(state *pb.BeaconState) (*pb.BeaconState, error) {
 	}
 	return state, nil
 }
+
+// ProcessValidatorRegistry computes and sets new validator registry fields,
+// reshuffles shard committees and returns the recomputed state.
+//
+// Spec pseudocode definition:
+//	Set state.validator_registry_latest_change_slot = state.slot.
+//	Set state.shard_committees_at_slots[:EPOCH_LENGTH] = state.shard_committees_at_slots[EPOCH_LENGTH:].
+//	Set state.shard_committees_at_slots[EPOCH_LENGTH:] =
+//  		get_new_shuffling(state.latest_randao_mixes[(state.slot - EPOCH_LENGTH) %
+//  		LATEST_RANDAO_MIXES_LENGTH], state.validator_registry, next_start_shard)
+//  		where next_start_shard = (state.shard_committees_at_slots[-1][-1].shard + 1) % SHARD_COUNT
+func ProcessValidatorRegistry(
+	state *pb.BeaconState) (*pb.BeaconState, error) {
+
+	epochLength := int(params.BeaconConfig().EpochLength)
+	randaoMixesLength := params.BeaconConfig().LatestRandaoMixesLength
+	shardCount := params.BeaconConfig().ShardCount
+
+	shardCommittees := state.ShardAndCommitteesAtSlots
+	lastSlot := len(shardCommittees) - 1
+	lastCommittee := len(shardCommittees[lastSlot].ArrayShardAndCommittee) - 1
+	nextStartShard := (shardCommittees[lastSlot].ArrayShardAndCommittee[lastCommittee].Shard + 1) %
+		shardCount
+
+	var randaoHash32 [32]byte
+	copy(randaoHash32[:], state.LatestRandaoMixesHash32S[(state.Slot-uint64(epochLength))%randaoMixesLength])
+
+	state.ValidatorRegistryLastChangeSlot = state.Slot
+	for i := 0; i < epochLength; i++ {
+		state.ShardAndCommitteesAtSlots[epochLength+i] = state.ShardAndCommitteesAtSlots[i]
```

Comment:
Shouldn't this be the other way around ? 
```go
state.ShardAndCommitteesAtSlots[i] = state.ShardAndCommitteesAtSlots[i+epochLength]
```
?

Reply:
Ah you are right. Good catch!